src/main/scala/mesosphere/marathon/storage/migration/MigrationTo18.scala

```diff
@@ -0,0 +1,227 @@
+      (__ \ "runSpecVersion").read[Timestamp] ~
+      (__ \ "state").read[ParsedValue[InstanceState]](instanceStateReads17) ~
+      (__ \ "unreachableStrategy").readNullable[raml.UnreachableStrategy] ~
+      (__ \ "reservation").readNullable[Reservation]
+    ) { (instanceId, agentInfo, tasksMap, runSpecVersion, state, ramlUnreachableStrategy, reservation) =>
+        val unreachableStrategy = ramlUnreachableStrategy.
+          map(Raml.fromRaml(_)).getOrElse(UnreachableStrategy.default())
+
+        if (List(state, tasksMap).exists(_.isModified)) {
+          val instance = new Instance(instanceId, Some(agentInfo), state.value, tasksMap.value, runSpecVersion, unreachableStrategy, reservation)
+          ParsedValue(instance, Modified)
+        } else {
+          val instance = new Instance(instanceId, Some(agentInfo), state.value, tasksMap.value, runSpecVersion, unreachableStrategy, reservation)
+          ParsedValue(instance, NotModified)
+        }
+
+      }
+  }
+
+  implicit val instanceResolver: IdResolver[Instance.Id, JsValue, String, ZkId] =
+    new IdResolver[Instance.Id, JsValue, String, ZkId] {
+      override def toStorageId(id: Id, version: Option[OffsetDateTime]): ZkId =
+        ZkId(category, id.idString, version)
+
+      override val category: String = "instance"
+
+      override def fromStorageId(key: ZkId): Id = Instance.Id.fromIdString(key.id)
+
+      override val hasVersions: Boolean = false
+
+      override def version(v: JsValue): OffsetDateTime = OffsetDateTime.MIN
+    }
+
+  implicit val instanceJsonUnmarshaller: Unmarshaller[ZkSerialized, JsValue] =
+    Unmarshaller.strict {
+      case ZkSerialized(byteString) =>
+        Json.parse(byteString.utf8String)
+    }
+
+  /**
+    * This function traverses all instances in ZK and sets the instance goal field.
+    */
+  def migrateInstanceConditions(instanceRepository: InstanceRepository, persistenceStore: PersistenceStore[_, _, _])(implicit mat: Materializer): Future[Done] = {
+
+    logger.info("Starting instance condition migration")
+
+    val countingSink: Sink[Done, NotUsed] = Sink.fold[Int, Done](0) { case (count, Done) => count + 1 }
+      .mapMaterializedValue { f =>
+        f.map(i => logger.info(s"$i instances migrated"))
+        NotUsed
+      }
+
+    val filterNotModified = Flow[ParsedValue[Instance]]
+      .mapConcat {
+        case ParsedValue(instance, Modified) =>
+          logger.info(s"${instance.instanceId} had `Created` condition, migration necessary")
+          List(instance)
+        case ParsedValue(instance, NotModified) =>
+          logger.info(s"${instance.instanceId} doesn't need to be migrated")
+          Nil
+      }
+
+    maybeStore(persistenceStore).map { store =>
+      instanceRepository
+        .ids()
+        .mapAsync(1) { instanceId =>
+          store.get[Instance.Id, JsValue](instanceId)
+        }
+        .via(parsingFlow)
+        .via(filterNotModified)
```

Comment:
isn't this too fancy? I find like we added so much code instead of:
- reading from repo
- `.filter(_.state.condition == Created)`
- `.copy(_.state.condition = Provisioned)`
-  persisting those back into repo

Reply:
we can't do `.filter(_.state.condition == Created)` because `Created` is removed from the codebase.